rfcs/001-assistant-prefix.md

```diff
@@ -0,0 +1,248 @@
+content to the chat template without a new turn:
+
+Example:
+```json
+"messages": [
+  { "role": "user", "content": "Tell me a joke." },
+  { "role": "assistant", "content": "Why did the chicken", "prefix": true }
+]
+```
+
+Example tokenized sequence for generation:
+
+```
+<begin_turn><role>user</role>
+Tell me a joke.
+
+<begin_turn><role>assistant</role>
+Why did the chicken
+```
+
+The server MUST respond with only the generated content, not including the prefix:
+
+Example:
+```json
+"messages": [
+  {
+    "role": "user",
+    "content": "Classify this color as either { type: light } or { type: dark }: blue"
+  },  
+  { "role": "assistant", "content": "```json\n", "prefix": true }
+],
+"stop": "\n```"
+```
+
+Example Response:
+```
+.choices[0].message.content = '{ "type": "dark" }'
+```
+
+### Token counting
+The server MUST count assistant prefix tokens as prompt tokens, not completion tokens,
+for the purposes of `max_tokens` and `usage`.
+
+### Errors
+
+If assistant prefixes are not supported for the requested model, the server MUST respond
+with `invalid_request_error`.
+
+## "prefix": false
+
+If the value of "prefix" on the trailing assistant message is false, the server MUST NOT
+render the message as a prefix. The message MUST be rendered as a separate turn. The
+message is not an assistant prefix.
+
+Example:
+```json
+"messages": [
+  { "role": "user", "content": "Tell me a joke." },
+  { "role": "assistant", "content": "Why did the chicken cross the road?", "prefix": false }
+]
+```
+
+Example tokenized sequence for generation:
+```
+<begin_turn><role>user</role>
+Tell me a joke.
+
+<begin_turn><role>assistant</role>
+Why did the chicken cross the road?
+
+<begin_turn><role>assistant</role>
+```
+
+### Errors
+
+If the requested model does not support multiple consecutive assistant messages, the
+server MUST return an `invalid_request_error`.
+
+## "prefix" unset
+
+If the prefix flag is unset, the result is implementation-defined. The server MAY treat
+the trailing assistant message as an assistant prefix, MAY render it as a new turn, or
+MAY return invalid_request_error.
+
+The server MAY use non-standard flags (such as `continue_final_message` on the request
+body) to decide how to handle a trailing assistant message with an unset prefix flag.
+
+*Note: This is left implementation-defined for backwards compatibility.*
+
+## Non-boolean "prefix" values
+
+Values of "prefix" that are not `true` or `false` are an error. The server MUST respond
+with an `invalid_request_error`. In the future, some non-boolean values may be
+supported, see *Unresolved Questions*.
+
+## "prefix" on user messages
+
+A "prefix" field on a user message is an error. The server MUST respond with an
+`invalid_request_error`.
+
+## "prefix" on non-trailing assistant messages
+
+A "prefix" field on a non-trailing assistant message MUST be ignored. The server MAY
+emit a warning.
```

Comment:
Reviewing this, I'm a bit unsure why we would have this be ignored, instead of generating invalid_request_error. It seems like a pretty clear developer coding error, and catching those earlier is probably good?

Reply:
Hm, good question. My thought was that people may have code like this:

```python
messages = [
  { "role": "user", "content": "blah blah blah" },
  { "role": "assistant", "content": "Certainly:", "prefix": True },
]
response_str = call_llm(messages)
messages[-1]["content"] = response

messages.append({ "role": "user", "content": "blah blah blah" })
response2_str = call_llm(messages)
```

And it might be nice to not require them to also clear the prefix flag. But I'm open to changing it if people think it'd be better to make this an error, it's not too onerous I don't think.